caas/kubernetes/provider/teardown.go

```diff
@@ -0,0 +1,263 @@
+			return labels
+		}
+		return nil
+	}
+	ensureResourcesDeletedFunc(ctx, labels, clk, wg, errChan,
+		func(labels map[string]string) error {
+			return k.deleteCustomResources(getLabels)
+		},
+		func(labels map[string]string) error {
+			_, err := k.listCustomResources(getLabels)
+			return err
+		},
+	)
+}
+
+func (k *kubernetesClient) deleteCustomResourceDefinitionsModelTeardown(
+	ctx context.Context,
+	labels map[string]string,
+	clk jujuclock.Clock,
+	wg *sync.WaitGroup,
+	errChan chan<- error,
+) {
+	ensureResourcesDeletedFunc(ctx, labels, clk, wg, errChan,
+		k.deleteCustomResourceDefinitions, func(labels map[string]string) error {
+			_, err := k.listCustomResourceDefinitions(labels)
+			return err
+		},
+	)
+}
+
+func (k *kubernetesClient) deleteMutatingWebhookConfigurationsModelTeardown(
+	ctx context.Context,
+	labels map[string]string,
+	clk jujuclock.Clock,
+	wg *sync.WaitGroup,
+	errChan chan<- error,
+) {
+	ensureResourcesDeletedFunc(ctx, labels, clk, wg, errChan,
+		k.deleteMutatingWebhookConfigurations, func(labels map[string]string) error {
+			_, err := k.listMutatingWebhookConfigurations(labels)
+			return err
+		},
+	)
+}
+
+func (k *kubernetesClient) deleteValidatingWebhookConfigurationsModelTeardown(
+	ctx context.Context,
+	labels map[string]string,
+	clk jujuclock.Clock,
+	wg *sync.WaitGroup,
+	errChan chan<- error,
+) {
+	ensureResourcesDeletedFunc(ctx, labels, clk, wg, errChan,
+		k.deleteValidatingWebhookConfigurations, func(labels map[string]string) error {
+			_, err := k.listValidatingWebhookConfigurations(labels)
+			return err
+		},
+	)
+}
+func (k *kubernetesClient) deleteStorageClassesModelTeardown(
+	ctx context.Context,
+	labels map[string]string,
+	clk jujuclock.Clock,
+	wg *sync.WaitGroup,
+	errChan chan<- error,
+) {
+	ensureResourcesDeletedFunc(ctx, labels, clk, wg, errChan,
+		k.deleteStorageClasses, func(labels map[string]string) error {
+			_, err := k.listStorageClasses(labels)
+			return err
+		},
+	)
+}
+
+type deleterChecker func(map[string]string) error
+
+func ensureResourcesDeletedFunc(
+	ctx context.Context,
+	labels map[string]string,
+	clk jujuclock.Clock,
+	wg *sync.WaitGroup,
+	errChan chan<- error,
+	deleter, checker deleterChecker,
+) {
+	defer wg.Done()
+
+	var err error
+	defer func() {
+		if err != nil {
+			select {
+			case errChan <- err:
+			default:
+			}
+		}
+	}()
+
+	if err = deleter(labels); err != nil {
+		if errors.IsNotFound(err) {
+			err = nil
+		}
+		return
+	}
+
+	interval := 1 * time.Second
+	ticker := clk.NewTimer(interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			err = errors.Trace(ctx.Err())
+			return
+		case <-ticker.Chan():
+			err = checker(labels)
+			if errors.IsNotFound(err) {
+				// Deleted already.
+				err = nil
+				return
+			}
+			if err != nil {
+				err = errors.Trace(err)
+				return
+			}
+		}
+		// Keep checking.
+		ticker.Reset(interval)
```

Comment:
I guess we don't use a timeout elsewhere in destroy-model
Not sure if we need one. Maybe useful to timeout after 10 or 15 minutes. The user can always be prompted to retry.
Not sure....

Reply:
Undertaker worker could pass timeout via dying chan then Destroy() will cancel the context.
Finally all these goroutines get aborted.

Reply:
Would it be better to use a timed context here over the ticker?